Sources/Signal.swift

```diff
@@ -38,132 +56,207 @@ public final class Signal<Value, Error: Swift.Error> {
 	///   - generator: A closure that accepts an implicitly created observer
 	///                that will act as an event emitter for the signal.
 	public init(_ generator: (Observer) -> Disposable?) {
-		state = Atomic(SignalState())
-
-		/// Holds the final signal state captured by an `interrupted` event. If it
-		/// is set, the Signal should interrupt as soon as possible. Implicitly
-		/// protected by `state` and `sendLock`.
-		var interruptedState: SignalState<Value, Error>? = nil
-
-		/// Used to track if the signal has terminated. Protected by `sendLock`.
-		var terminated = false
-
-		/// Used to ensure that events are serialized during delivery to observers.
-		let sendLock = NSLock()
-		sendLock.name = "org.reactivecocoa.ReactiveSwift.Signal"
+		state = .alive(SignalStateSnapshot())
+		updateLock = NSLock()
+		updateLock.name = "org.reactivecocoa.ReactiveSwift.Signal.updateLock"
+		sendLock = NSLock()
+		sendLock.name = "org.reactivecocoa.ReactiveSwift.Signal.sendLock"
 
 		let observer = Observer { [weak self] event in
 			guard let signal = self else {
 				return
 			}
 
-			@inline(__always)
-			func interrupt(_ observers: Bag<Observer>) {
-				for observer in observers {
-					observer.sendInterrupted()
-				}
-				terminated = true
-				interruptedState = nil
-			}
-
-			if case .interrupted = event {
-				// Recursive events are generally disallowed. But `interrupted` is kind
-				// of a special snowflake, since it can inadvertently be sent by
-				// downstream consumers.
+			// Thread Safety Notes on `Signal.state`.
+			//
+			// - Check if the signal is at a specific state.
+			//
+			//   Read directly.
+			//
+			// - Extract the snapshot of a signal that is alive, and deliver events
+			//   using the snapshot.
+			//
+			//   `sendLock` must be acquired.
+			//
+			// - Replace the snapshot of a signal that is alive.
+			//   (e.g. observers bag manipulation)
+			//
+			//   `updateLock` must be acquired.
+			//
+			// - Transit from `alive` to `terminating`.
```

Comment:
I guess this needs a rename too, hmm?
https://github.com/ReactiveCocoa/ReactiveSwift/blob/master/Sources/Atomic.swift#L27